### ISSUE TITLE
How do I register and unregister a speaker Proxy, where is  iAppCallBackAPI ?

### ISSUE THREAD

**Issue by dcorsus** [HomeSeer/Plugin-SDK#3]
These are the HS3 equivalent methods RegisterProxySpeakPlug and UnRegisterProxySpeakPlug. There is also a call back involved which was part of  iAppCallBackAPI.
Without this, I can't do TTS ... but you already knew that :-)

**Comment by spudwebb** [comment]
> Without this, I can't do TTS ... but you already knew that :-)

Actually you can, there is the API used by the HomeSeer Speaker Client (IFromSpeaker / ISpeechAPI), I recently switched my HS3 Chromecast plugin from SpeakProxy to this API. It works well and it has the added benefit that your speakers directly appear in the Speaker Client List for a Speak event action. 
We need to discuss with @rjhdakota if this is the recommended way of doing it from now on.

This API is in the PluginSdk (HomeSeer.PluginSdk.Speech), but I haven't tested it with a HS4 plugin yet.

**Comment by dcorsus** [comment]
@spudwebb 
That is an interesting suggestion. Rich very recently suggested the same when I asked for a capability to add speakerclients to the list of select-able speaker clients. He suggest to use the HSTouch SpeakerClient class as opposed to having the capability to add static entries to the list of select-able speaker client lists. I'm still worried, despite Rich telling me not to, to open more interfaces. For example, in my PI, a user can create many different groupings or settings of Sonos players for announcements. So in HS2 and HS3, the way I passed that information from the user created event into my PI was to use a 'fake speakerclient name" which was constructed with a prefix $SONOS$ and a "grouping name" say $Entire House$, so it would make up $SONOS$Entire House$. So I can have many of these and If I now need to instantiate a call, which I suspect may open another UDP/TCP connection per entry to fake these "speaker clients", that makes me nervous. The proxy as just one single interface. Maybe I should study the speaker Client class, if I can instantiate only one and have a way to pass the information I need (the linkgroup name), that would work.
Thoughts or experience with this?

**Comment by jldubz** [comment]
iAppCallbackAPI has been deprecated and its functionality was rolled into iHsController.

We will be discussing the speech API internally to determine exactly what the recommended approach is going to be and will update you when we can provide clearer instructions.

Thanks!

**Comment by dcorsus** [comment]
@jldubz @spudwebb 
Is there anything happening on this front?
I tried to go the "API used by the HomeSeer Speaker Client (IFromSpeaker / ISpeechAPI)" way but a bit stuck on how to create a connection. How do I instantiate the class IspeechAPI to issue a ConnectLocal call? If this class is already instantiated, how do I get a handle to it?
Help!

**Comment by jldubz** [comment]
@dcorsus,
Thanks for your patience on this.  Would you be able to send me an example of how you had it implemented in HS3?  The internal plugin samples we have don't include a robust implementation, and I want to make sure the resources available for you in this SDK are sufficient for your needs.

**Comment by dcorsus** [comment]
I had it implemented very straight forward as per the HS3 guidelines.

I see in the main procedure that 2 TCP endpoints are created, sure this is straight from the HS3 sample PI.

        client = ScsServiceClientBuilder.CreateClient(Of IHSApplication)(New ScsTcpEndPoint(sIp, 10400), gAppAPI)
        clientCallback = ScsServiceClientBuilder.CreateClient(Of IAppCallbackAPI)(New ScsTcpEndPoint(sIp, 10400), gAppAPI)

The clientCallback of type iAppCallbaklAPI had a method clientCallback.ServiceProxy that would return a handle of type iAppCallback.

This handle was used for things like:
RegisterConfigLink
RegisterProxySpeakPlug (PIName, PIInstance)
UnRegisterProxySpeakPlug
..

Then you add this method to your PI and each time ANY PI or HS invokes the Speak method, your PI gets a call back through this method

  Public Sub SpeakIn(ByVal device As Integer, ByVal text As String, ByVal wait As Boolean, ByVal host As String) Implements HomeSeerAPI.IPlugInAPI.SpeakIn

I'm not sure this is what you are asking, but I used the plain vanilla HS3 suggested implementation.

Now the "cheat" I would do, is to ask my PI users, when they created an event to "add a speakerclient" MANUALLY and give it a very specific name, example: $SONOS$All Rooms$.

When this action was invoked the SpeakIn method listed above was called (back) and the host parameter would be equal to $SONOS$All Rooms$. So the PI would use that info to take action or not.

Some users complained they had to remember and understand the $SONOS$All Room$ syntax and complained that for each Speak event they created, they had to do all the manual steps to create a fake speaker client again and again.

So the question for HS4 implementation was whether the PI itself could populate these fake speakerclient entries in the selection menu when users create speak events. Rich pitched in and suggested I used the stripped down code they use in HSTouch, and apparently Spud, who I'm sure is reading this, is doing the same. This is implementing the IFromSpeaker class. I applied this in HS4 and I have this running. 

What I don't like so much about it is that IFromSpeaker has a ton of mandated inherited methods (~30) so is isn't quite user friendly. I guess one way is to define a class in the SDK that hides all non-essential methods.

Second thing which is a bit more problematic is that I haven't found a way to pass information. If you go back to my example $SONOS$All Rooms$, you can see 2 parts. The $SONOS$ part is just to help the HS3 Proxy to pick out those proxy calls that are meant for the Sonos PI and ignore all else. The second part however, is crucial user information. It tells the Sonos PI what to do, ie which players should participate in the announcement.

So there are a few alternatives here for me:
a/ create a specific IFromSpeaker speaker client instance for each configuration I have. The drawback is that each will have an active TCP connection, so something not very efficient for what I need.
b/ Stay with the HS3 implementation, meaning no implementation of IFromSpeaker done by PI authors, make available that one callback function SpeakIn. Advantage is that migration from HS3 to HS4 doesn't break existing events and is quite efficient but had plenty of issues in the past I think with threading or blocking of HS.
c/ Allow the PI authors to add entries to the dropdown list in the speakevent settings screen, instead of the users each time having to add a manual entry. Even if that info was cached it would already be an improvement. However last I looked the function to add manually a fake speaker client is gone, so at this point, I guess the only way is option a above. 

Perhaps you can dumb that down to the bare minimum with an abstractclass and allow non-essential methods to be overwritten, maybe one common TCP connection as opposed to mutitple?

Clear as mud?

**Comment by spudwebb** [comment]
> a/ create a specific IFromSpeaker speaker client instance for each configuration I have. The drawback is that each will have an active TCP connection, so something not very efficient for what I need.

This is what I do for the Chromecast plugin, one IFromSpeaker speaker client instance for each speaker.
There is another drawback with IFromSpeaker, it is that if in the network settings "No Password Required for Local (same subnet) Login" is unchecked, then you have to pass a valid username/password to be able to connect in ISpeechAPI.Connect(), which seems a little weird because plugin don't have to use any user/password to connect to HS

**Comment by jldubz** [comment]
Working on an integrated implementation of all of this for next week's v1.0.7 release using AbstractSpeakerClient

**Comment by jldubz** [comment]
This is being delayed until post beta release.

**Comment by jldubz** [comment]
Tracked by PSDK-5.  Investigating it for v1.3.0

**Comment by rjhelmke** [comment]
Is this still an issue for you? I think you have this working now?
